### PR TITLE
[colocated-join] Add Support for Deterministic Segment Assignment Strategy

### PR DESCRIPTION
**Pull request by ankitsultana**
Will use this PR to discuss some questions offline and update details here afterwards.

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/OfflineSegmentAssignment.java [hunk 1]
```diff
@@ -88,7 +89,12 @@ public void init(HelixManager helixManager, TableConfig tableConfig) {
     if (_partitionColumn == null) {
       LOGGER.info("Initialized OfflineSegmentAssignment with replication: {} without partition column for table: {} ",
           _replication, _offlineTableName);
+      _useDeterministicSegmentAssignment = false;
     } else {
+      // TODO: Where do I keep the segment assignment strategy
```
Comment (ankitsultana) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/OfflineSegmentAssignment.java:94]: Where's the right place to keep the segment assignment strategy constant?

I saw CommonConstants has a SegmentAssignmentStrategyType enum, but it's not getting used anywhere.

**Review by Jackie-Jiang — COMMENTED** [review]
This has the same behavior as setting `numInstancesPerPartition` to 1, and `numPartitions` to the partitions for the table.
Currently we don't support explicitly setting partitions for `RealtimeSegmentAssignment`, so we should fix that instead of introducing this new strategy

**Comment by yupeng9** [comment]
> This has the same behavior as setting `numInstancesPerPartition` to 1, and `numPartitions` to the partitions for the table. Currently we don't support explicitly setting partitions for `RealtimeSegmentAssignment`, so we should fix that instead of introducing this new strategy

I feel this is not exactly the same as `numInstancesPerPartition=1`, instead `numInstancesPerPartition` shall be the same as the number of replicas. In fact, the strategy is somewhat deterministic as the added stragy name suggests. And the main purpose is to make the assignment declarative and reduce the operational cost, e.g. we dont need to update `numInstancesPerPartition` when we increase `numPartitions` to scale out the table etc

**Comment by Jackie-Jiang** [comment]
> > This has the same behavior as setting `numInstancesPerPartition` to 1, and `numPartitions` to the partitions for the table. Currently we don't support explicitly setting partitions for `RealtimeSegmentAssignment`, so we should fix that instead of introducing this new strategy
> 
> I feel this is not exactly the same as `numInstancesPerPartition=1`, instead `numInstancesPerPartition` shall be the same as the number of replicas. In fact, the strategy is somewhat deterministic as the added stragy name suggests. And the main purpose is to make the assignment declarative and reduce the operational cost, e.g. we dont need to update `numInstancesPerPartition` when we increase `numPartitions` to scale out the table etc

This new strategy is trying to put all segments from the same segment partition to the same server, which is exactly the same as what `numInstancesPerPartition = 1` is doing (and it is also deterministic).
In order to achieve this, you should use the partition based replica-group assignment, and always put `numInstancesPerPartition = 1`. You may increase `numPartitions` if needed (should be very rare because it should be the same as table partitions, and changing it requires re-partitioning all the segments). Noted that `numPartitions` can be larger than the total instances within a replica-group. E.g. with 9 servers, you can have 3 replica-groups (3 servers per replica), then 18 partitions, and each server will host 6 partitions. You may scale it up to 18 servers, and run a rebalance without changing `numInstancesPerPartition` or `numPartitions`. After rebalancing, each server will host 3 partitions.

**Comment by yupeng9** [comment]
> 

I see whaat you mean. Then that's a reasonable suggestion. But I hope we can find a way to enforce this as it's not so easy for users to understand the details to setup colocation correctly. I feel either some alias, or some validation in the table config is needed.

**Comment by Jackie-Jiang** [comment]
@yupeng9 True. I discussed this with @ankitsultana and suggested that we may introduce a flag to mark colocated tables, then we can check the table config to enforce `numInstancesPerPartition=1`

**Comment by ankitsultana** [comment]
> Currently we don't support explicitly setting partitions for RealtimeSegmentAssignment, so we should fix that instead of introducing this new strategy

Yeah I'll start working on it. I am not so sure about using numInstancesPerPartition=1, but at the same time I think I need to take a holistic look at the colocation setup once we have done some PoCs internally (ETA: 2 weeks from now).

**Pull request closed without merging**